components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/action/PreIssueAccessTokenRequestBuilder.java

```diff
@@ -0,0 +1,330 @@
+            return tokenMessageContext.getAudiences();
+        } else {
+            return OAuth2Util.getOIDCAudience(oAuthAppDO.getOauthConsumerKey(), oAuthAppDO);
+        }
+    }
+
+    private void handleStandardClaims(OAuthTokenReqMessageContext tokenMessageContext, String tokenType,
+                                      String issuer, List<String> audience, AccessToken.Builder accessTokenBuilder) {
+
+        accessTokenBuilder.tokenType(tokenType)
+                .addClaim(AccessToken.ClaimNames.ISS.getName(), issuer)
+                .addClaim(AccessToken.ClaimNames.CLIENT_ID.getName(),
+                        tokenMessageContext.getOauth2AccessTokenReqDTO().getClientId())
+                .addClaim(AccessToken.ClaimNames.AUTHORIZED_USER_TYPE.getName(),
+                        String.valueOf(tokenMessageContext.getProperty(OAuthConstants.UserType.USER_TYPE)))
+                .addClaim(AccessToken.ClaimNames.EXPIRES_IN.getName(),
+                        tokenMessageContext.getValidityPeriod() / 1000)
+                .addClaim(AccessToken.ClaimNames.AUD.getName(), audience)
+                .scopes(Arrays.asList(tokenMessageContext.getScope()));
+    }
+
+    private void handleSubjectClaim(AuthenticatedUser authorizedUser, OAuthAppDO oAuthAppDO,
+                                    AccessToken.Builder accessTokenBuilder) throws IdentityOAuth2Exception {
+
+        String sub = authorizedUser.getAuthenticatedSubjectIdentifier();
+        if (OAuth2Util.isPairwiseSubEnabledForAccessTokens()) {
+            sub = OIDCClaimUtil.getSubjectClaim(sub, oAuthAppDO);
+            accessTokenBuilder.addClaim(AccessToken.ClaimNames.SUBJECT_TYPE.getName(),
+                    OIDCClaimUtil.getSubjectType(oAuthAppDO).getValue());
+        }
+        accessTokenBuilder.addClaim(AccessToken.ClaimNames.SUB.getName(), sub);
+    }
+
+    private Map<String, Object> getAdditionalClaimsToAddToToken(OAuthTokenReqMessageContext tokenMessageContext)
+            throws ActionExecutionRequestBuilderException {
+         /*
+         Directly return custom claims if pre-issue access token actions have been executed.
+         This is to ensure that the custom claims added are incorporated in the refresh token flow.
+         Moreover, this execution expects that the claim handlers executed at the token issuance flow
+         does not incorporate any additional custom rules based on refresh grant.
+         */
+        if (tokenMessageContext.isPreIssueAccessTokenActionsExecuted()) {
+            return tokenMessageContext.getAdditionalAccessTokenClaims();
+        }
+
+        try {
+            CustomClaimsCallbackHandler claimsCallBackHandler =
+                    OAuthServerConfiguration.getInstance().getOpenIDConnectCustomClaimsCallbackHandler();
+            JWTClaimsSet claimsSet =
+                    claimsCallBackHandler.handleCustomClaims(new JWTClaimsSet.Builder(), tokenMessageContext);
```

Comment:
There are a couple of claims being added before calling the claims handler in the previous flow. Don't we need those claims?

Reply:
Only non standard or claims that are not specially handled are treated this way (Ideally, user's org, accessing org should be default incorporated). Will refer as special claims going forward. What this means is claims that the product defines to be incorporated to the access token without letting the developer to pick and choose from app configuration or incorporate from the extension. 
In the token processor I have a logic to filter out these special claims from the set of claims that developer has control of
https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2528/files#diff-ba5e6b8dbaba53f900aa6916029a107db0bee843cdcc3f5ff0d6e6a99d4caf15R166-R172

Hope this answers.
However, as of now, we don't have a common way of differentiating the claims in the access token that system incorporates or incorporated by the developer over some configuration or extension. Therefore, this code block handles such specially and that code segment is duplicated in JWTTokenIssuer as well

Reply:
ack